File: php/src/RecommendAIClient.php
<?php

declare(strict_types=1);

namespace RecommendAI;

use GuzzleHttp\Client as GuzzleClient;
use GuzzleHttp\Exception\ClientException;
use GuzzleHttp\Exception\ServerException as GuzzleServerException;

/**
 * Main entry point for the RecSys.AI PHP SDK.
 */
final class RecommendAIClient
{
    private readonly GuzzleClient $http;
    private readonly RecommendationsResource $recommendationsResource;
    private readonly UsersResource $usersResource;
    private readonly ItemsResource $itemsResource;
    private readonly InteractionsResource $interactionsResource;

    public function __construct(
        string        $apiKey,
        string        $baseUrl = 'http://localhost:8080',
        int           $timeout = 30,
        ?GuzzleClient $httpClient = null,
    ) {
        $this->http = $httpClient ?? new GuzzleClient([
            'base_uri' => rtrim($baseUrl, '/') . '/',
            'timeout'  => $timeout,
            'headers'  => [
                'Authorization' => "Bearer {$apiKey}",
                'Content-Type'  => 'application/json',
                'Accept'        => 'application/json',
                'User-Agent'    => 'recommendai-php/1.0.0',
            ],
        ]);

        $this->recommendationsResource = new RecommendationsResource($this->http);
        $this->usersResource           = new UsersResource($this->http);
        $this->itemsResource           = new ItemsResource($this->http);
        $this->interactionsResource    = new InteractionsResource($this->http);
    }

    public function recommendations(): RecommendationsResource
    {
        return $this->recommendationsResource;
    }

    public function users(): UsersResource
    {
        return $this->usersResource;
    }

    public function items(): ItemsResource
    {
        return $this->itemsResource;
    }

    public function interactions(): InteractionsResource
    {
        return $this->interactionsResource;
    }

    public function ping(): bool
    {
        try {
            $this->http->get('health');
            return true;
        } catch (\Throwable) {
            return false;
        }
    }

    /**
     * Map a Guzzle HTTP error to a typed RecommendAI exception.
     *
     * @internal used by resource classes
     */
    public static function mapError(\Throwable $e): RecommendAIException
    {
        if ($e instanceof ClientException) {
            $status = $e->getResponse()->getStatusCode();
            $body   = (string) $e->getResponse()->getBody();
            $data   = json_decode($body, true);
            $msg    = is_array($data) ? ($data['detail'] ?? "HTTP {$status}") : "HTTP {$status}";
            return match ($status) {
                401     => new AuthenticationException($msg),
                404     => new NotFoundException($msg),
                429     => new RateLimitException($msg),
                400, 422 => new ValidationException($msg, $status),
                default => new RecommendAIException($msg, $status),
            };
        }
        if ($e instanceof GuzzleServerException) {
            $status = $e->getResponse()->getStatusCode();
            $body   = (string) $e->getResponse()->getBody();
            $data   = json_decode($body, true);
            $msg    = is_array($data) ? ($data['detail'] ?? "HTTP {$status}") : "HTTP {$status}";
            return new ServerException($msg, $status);
        }
        return new RecommendAIException($e->getMessage());
    }
}

File: dotnet/src/RecommendAIClient.cs
using System.Net.Http.Headers;
using System.Runtime.CompilerServices;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

[assembly: InternalsVisibleTo("RecommendAI.Tests")]

namespace RecommendAI;

/// <summary>
/// Official .NET client for the RecSys.AI personalised-recommendation platform.
/// </summary>
public sealed class RecommendAIClient : IDisposable
{
    internal HttpClient Http;

    private static readonly JsonSerializerOptions JsonOpts = new()
    {
        PropertyNameCaseInsensitive = true,
        DefaultIgnoreCondition      = JsonIgnoreCondition.WhenWritingNull,
        Converters = { new JsonStringEnumConverter() },
    };

    /// <summary>Access recommendations endpoints.</summary>
    public RecommendationsResource Recommendations { get; }
    /// <summary>Access user management endpoints.</summary>
    public UsersResource Users { get; }
    /// <summary>Access item catalogue endpoints.</summary>
    public ItemsResource Items { get; }
    /// <summary>Access interaction recording endpoints.</summary>
    public InteractionsResource Interactions { get; }

    /// <summary>Creates a new client.</summary>
    /// <param name="apiKey">Your RecSys.AI API key.</param>
    /// <param name="baseUrl">Override the default API base URL.</param>
    /// <param name="timeout">HTTP request timeout (default 30 s).</param>
    public RecommendAIClient(
        string apiKey,
        string baseUrl = "http://localhost:8080",
        TimeSpan? timeout = null)
    {
        Http = new HttpClient
        {
            BaseAddress = new Uri(baseUrl.TrimEnd('/') + "/"),
            Timeout     = timeout ?? TimeSpan.FromSeconds(30),
        };
        Http.DefaultRequestHeaders.Authorization =
            new AuthenticationHeaderValue("Bearer", apiKey);
        Http.DefaultRequestHeaders.UserAgent.ParseAdd("recommendai-dotnet/1.0.0");

        Recommendations = new RecommendationsResource(this);
        Users           = new UsersResource(this);
        Items           = new ItemsResource(this);
        Interactions    = new InteractionsResource(this);
    }

    /// <summary>Internal constructor for testing — injects a custom message handler.</summary>
    internal RecommendAIClient(string apiKey, HttpMessageHandler handler)
    {
        Http = new HttpClient(handler)
        {
            BaseAddress = new Uri("http://localhost:8080/"),
        };
        Http.DefaultRequestHeaders.Authorization =
            new AuthenticationHeaderValue("Bearer", apiKey);
        Http.DefaultRequestHeaders.UserAgent.ParseAdd("recommendai-dotnet/1.0.0");

        Recommendations = new RecommendationsResource(this);
        Users           = new UsersResource(this);
        Items           = new ItemsResource(this);
        Interactions    = new InteractionsResource(this);
    }

    // ── Internal helpers ──────────────────────────────────────────────────────

    internal static StringContent ToJson(object payload)
        => new(JsonSerializer.Serialize(payload, JsonOpts), Encoding.UTF8, "application/json");

    internal async Task<T> SendAsync<T>(HttpRequestMessage request)
    {
        var response = await Http.SendAsync(request).ConfigureAwait(false);
        var body     = await response.Content.ReadAsStringAsync().ConfigureAwait(false);

        if (!response.IsSuccessStatusCode)
            throw BuildException((int)response.StatusCode, body);

        if (response.StatusCode == System.Net.HttpStatusCode.NoContent || string.IsNullOrWhiteSpace(body))
            return default!;

        return JsonSerializer.Deserialize<T>(body, JsonOpts)
               ?? throw new RecommendAIException("Empty response body");
    }

    internal async Task SendAsync(HttpRequestMessage request)
    {
        var response = await Http.SendAsync(request).ConfigureAwait(false);
        if (!response.IsSuccessStatusCode)
        {
            var body = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
            throw BuildException((int)response.StatusCode, body);
        }
    }

    private static RecommendAIException BuildException(int status, string body)
    {
        string detail = ExtractDetail(body) ?? $"HTTP {status}";
        return status switch
        {
            401                  => new AuthenticationException(detail),
            404                  => new NotFoundException(detail),
            400 or 422           => new ValidationException(detail, status),
            429                  => new RateLimitException(detail),
            >= 500 and <= 599    => new ServerException(detail, status),
            _                    => new RecommendAIException(detail, status),
        };
    }

    private static string? ExtractDetail(string body)
    {
        try
        {
            using var doc = JsonDocument.Parse(body);
            if (doc.RootElement.TryGetProperty("detail", out var d))
                return d.GetString();
        }
        catch { /* fall through */ }
        return null;
    }

    /// <summary>Returns true if the API is reachable.</summary>
    public async Task<bool> PingAsync()
    {
        try
        {
            var req      = new HttpRequestMessage(HttpMethod.Get, "health");
            var response = await Http.SendAsync(req).ConfigureAwait(false);
            return response.IsSuccessStatusCode;
        }
        catch { return false; }
    }

    /// <inheritdoc />
    public void Dispose() => Http.Dispose();
}
